File: frontend/src/types/quickbooks.ts
/**
 * QuickBooks integration TypeScript types.
 *
 * Maps to backend schemas in backend/app/schemas/quickbooks.py
 */

// ---------------------------------------------------------------------------
// Connection
// ---------------------------------------------------------------------------

export interface QBConnectionStatus {
  is_connected: boolean;
  realm_id?: string;
  company_name?: string;
  connected_at?: string;
  last_sync_at?: string;
  last_sync_status?: string;
  token_expires_at?: string;
}

// ---------------------------------------------------------------------------
// Templates
// ---------------------------------------------------------------------------

export interface QBTemplateMappingDef {
  mapping_type: string;
  name: string;
  account_type: string;
  account_sub_type: string;
  is_default: boolean;
  description: string;
}

export interface QBTemplateInfo {
  template_name: string;
  name: string;
  description: string;
  mapping_count: number;
  mappings: QBTemplateMappingDef[];
}

export type TemplateCategory =
  | "all"
  | "pakistani"
  | "international"
  | "format"
  | "specialty"
  | "niche";

export const TEMPLATE_CATEGORIES: Record<string, TemplateCategory> = {
  // Pakistani (8)
  pakistani_restaurant: "pakistani",
  pakistani_bbq_specialist: "pakistani",
  biryani_house: "pakistani",
  pakistani_street_food: "pakistani",
  nihari_paye_house: "pakistani",
  pakistani_sweets_bakery: "pakistani",
  karachi_seafood: "pakistani",
  lahore_food_street: "pakistani",
  // International (8)
  international_restaurant: "international",
  chinese_restaurant: "international",
  pizza_chain: "international",
  burger_joint: "international",
  steakhouse: "international",
  japanese_sushi: "international",
  thai_restaurant: "international",
  italian_restaurant: "international",
  // Format (10)
  qsr: "format",
  cafe: "format",
  fine_dining: "format",
  buffet_restaurant: "format",
  food_court_vendor: "format",
  cloud_kitchen: "format",
  food_truck: "format",
  catering_company: "format",
  hotel_restaurant: "format",
  bar_lounge: "format",
  // Specialty (8)
  juice_bar: "specialty",
  ice_cream_parlor: "specialty",
  bakery_wholesale: "specialty",
  breakfast_spot: "specialty",
  dessert_parlor: "specialty",
  tea_house: "specialty",
  shawarma_wrap_shop: "specialty",
  fried_chicken_chain: "specialty",
  // Niche (6)
  electronics_food: "niche",
  vegan_vegetarian: "niche",
  organic_farm_to_table: "niche",
  dark_kitchen: "niche",
  diner: "niche",
  roti_canteen: "niche",
};

export const CATEGORY_LABELS: Record<TemplateCategory, string> = {
  all: "All",
  pakistani: "Pakistani",
  international: "International",
  format: "Format / Model",
  specialty: "Specialty",
  niche: "Niche",
};

// ---------------------------------------------------------------------------
// Shared Constants
// ---------------------------------------------------------------------------

/** Canonical mapping type labels — single source of truth for all QB tabs */
export const MAPPING_TYPE_LABELS: Record<string, string> = {
  income: "Income",
  cogs: "Cost of Goods Sold",
  tax_payable: "Tax Payable",
  bank: "Bank / Cash",
  expense: "Expenses",
  discount: "Discounts",
  rounding: "Rounding",
  cash_over_short: "Cash Over/Short",
  tips: "Tips",
  service_charge: "Service Charge",
  delivery_fee: "Delivery Fee",
  foodpanda_commission: "Platform Commission",
  gift_card_liability: "Gift Cards",
  other_current_liability: "Other Liabilities",
};

// ---------------------------------------------------------------------------
// Account Mappings
// ---------------------------------------------------------------------------

export interface QBAccountMapping {
  id: string;
  connection_id: string;
  mapping_type: string;
  pos_reference_id?: string;
  pos_reference_type?: string;
  pos_reference_name?: string;
  qb_account_id: string;
  qb_account_name: string;
  qb_account_type: string;
  qb_account_sub_type?: string;
  is_default: boolean;
  is_auto_created: boolean;
  created_at: string;
}

export interface QBAccountMappingCreate {
  mapping_type: string;
  qb_account_id: string;
  qb_account_name: string;
  qb_account_type: string;
  qb_account_sub_type?: string;
  pos_reference_id?: string;
  pos_reference_type?: string;
  pos_reference_name?: string;
  is_default?: boolean;
}

// ---------------------------------------------------------------------------
// Smart Defaults
// ---------------------------------------------------------------------------

export interface QBSmartDefaultsResult {
  accounts_created: number;
  mappings_created: number;
  mappings_skipped: number;
  details: string[];
}

// ---------------------------------------------------------------------------
// Sync
// ---------------------------------------------------------------------------

export interface QBSyncStats {
  total_synced: number;
  last_24h_synced: number;
  last_24h_failed: number;
  pending_jobs: number;
  failed_jobs: number;
  dead_letter_jobs: number;
  last_sync_at?: string;
  sync_by_type: Record<string, number>;
}

export interface QBSyncJob {
  id: string;
  job_type: string;
  entity_type: string;
  entity_id?: string;
  priority: number;
  status: string;
  error_message?: string;
  retry_count: number;
  created_at: string;
  completed_at?: string;
  processing_duration_ms?: number;
}

export interface QBSyncLog {
  id: string;
  sync_type: string;
  pos_entity_type?: string;
  pos_entity_id?: string;
  qb_entity_type?: string;
  qb_entity_id?: string;
  action: string;
  status: string;
  error_message?: string;
  error_code?: string;
  duration_ms?: number;
  qb_doc_number?: string;
  amount_paisa?: number;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Preview / Simulation
// ---------------------------------------------------------------------------

export interface QBPreviewResponse {
  template_name: string;
  template_display_name: string;
  order_number: string;
  order_type: string;
  qb_entity_type: string;
  payload: Record<string, unknown>;
  mappings_used: QBTemplateMappingDef[];
}

// ---------------------------------------------------------------------------
// Diagnostic & Onboarding Tool
// ---------------------------------------------------------------------------

export interface QBMatchSignals {
  exact: number;
  jaccard: number;
  synonym: number;
  type_match: number;
  substring: number;
}

export interface QBMatchCandidate {
  qb_account_id: string;
  qb_account_name: string;
  qb_account_type: string;
  qb_account_sub_type?: string;
  fully_qualified_name?: string;
  active: boolean;
  score: number;
  signals: QBMatchSignals;
  confidence: "high" | "medium" | "low";
}

export interface QBDiagnosticItem {
  mapping_type: string;
  template_account_name: string;
  template_account_type: string;
  template_account_sub_type?: string;
  template_description: string;
  is_default: boolean;
  status: "matched" | "candidates" | "unmatched";
  best_match: QBMatchCandidate | null;
  candidates: QBMatchCandidate[];
  decision: "use_existing" | "create_new" | "skip" | null;
  decision_account_id: string | null;
  decision_account_name: string | null;
}

export interface QBUnmappedAccount {
  qb_account_id: string;
  qb_account_name: string;
  qb_account_type: string;
  qb_account_sub_type?: string;
  fully_qualified_name?: string;
  active: boolean;
  suggested_mapping_type: string | null;
}

export interface QBDecisionSummary {
  use_existing: number;
  create_new: number;
  skip: number;
  pending: number;
  ready_to_apply: boolean;
}

export interface QBDiagnosticReport {
  id: string;
  template_key: string;
  template_name: string;
  created_at: string;
  fixture_name: string | null;
  is_live: boolean;
  total_template_mappings: number;
  total_qb_accounts: number;
  matched: number;
  candidates: number;
  unmatched: number;
  coverage_pct: number;
  health_grade: "A" | "B" | "C" | "F";
  summary: string;
  items: QBDiagnosticItem[];
  unmapped_qb_accounts: QBUnmappedAccount[];
  decision_summary?: QBDecisionSummary;
  apply_result?: Record<string, unknown>;
  applied_at?: string;
}

export interface QBDiagnosticReportSummary {
  id: string;
  template_key: string;
  template_name: string;
  created_at: string;
  health_grade: string;
  matched: number;
  candidates: number;
  unmatched: number;
  total_template_mappings: number;
  coverage_pct: number;
  is_live: boolean;
  fixture_name: string | null;
}

export interface QBDiagnosticDecision {
  index: number;
  decision: "use_existing" | "create_new" | "skip";
  qb_account_id?: string;
  qb_account_name?: string;
}

export interface QBDiagnosticApplyResult {
  accounts_created: number;
  mappings_created: number;
  skipped: number;
  errors: string[];
  details: string[];
  report_id: string;
}

export interface QBHealthCheckDetail {
  mapping_id: string;
  mapping_type: string;
  qb_account_id: string;
  qb_account_name: string;
  status: "healthy" | "warning" | "critical";
  issue: string | null;
  message: string;
  current_name: string | null;
}

export interface QBHealthCheckResult {
  grade: "A" | "B" | "C" | "F";
  total_mappings: number;
  healthy: number;
  warnings: number;
  critical: number;
  checked_at: string;
  details: QBHealthCheckDetail[];
}

export interface QBTestFixture {
  name: string;
  description: string;
  account_count: number;
}

File: backend/app/schemas/quickbooks.py
"""Pydantic v2 schemas for QuickBooks Online integration (Attempt 2).

Client-centric approach — no templates. POS declares what it needs,
fuzzy-matches against partner's actual QB Chart of Accounts.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# OAuth Flow
# ---------------------------------------------------------------------------


class QBConnectURL(BaseModel):
    auth_url: str
    state: str


class QBConnectionStatus(BaseModel):
    is_connected: bool
    connection_type: str | None = None  # "online" | "desktop"
    realm_id: str | None = None  # QB Online only
    company_name: str | None = None
    connected_at: datetime | None | str = None
    last_sync_at: datetime | None = None
    last_sync_status: str | None = None
    token_expires_at: datetime | None = None  # QB Online only
    # QB Desktop fields
    qbwc_username: str | None = None
    qb_desktop_version: str | None = None
    last_qbwc_poll_at: datetime | None = None
    model_config = {"from_attributes": True}


class QBDisconnectResponse(BaseModel):
    message: str
    realm_id: str | None = None


class QBCompanyInfo(BaseModel):
    company_name: str
    legal_name: str | None = None
    country: str | None = None
    fiscal_year_start: str | None = None
    industry_type: str | None = None
    currency: str | None = None


# ---------------------------------------------------------------------------
# Account Mappings
# ---------------------------------------------------------------------------


class QBAccountMappingBase(BaseModel):
    mapping_type: str = Field(..., min_length=1, max_length=50)
    pos_reference_id: uuid.UUID | None = None
    pos_reference_type: str | None = Field(None, max_length=50)
    pos_reference_name: str | None = Field(None, max_length=255)
    qb_account_id: str = Field(..., min_length=1, max_length=100)
    qb_account_name: str = Field(..., min_length=1, max_length=255)
    qb_account_type: str = Field(..., min_length=1, max_length=100)
    qb_account_sub_type: str | None = Field(None, max_length=100)
    is_default: bool = False


class QBAccountMappingCreate(QBAccountMappingBase):
    pass


class QBAccountMappingUpdate(BaseModel):
    qb_account_id: str | None = Field(None, min_length=1, max_length=100)
    qb_account_name: str | None = Field(None, min_length=1, max_length=255)
    qb_account_type: str | None = Field(None, min_length=1, max_length=100)
    qb_account_sub_type: str | None = Field(None, max_length=100)
    is_default: bool | None = None


class QBAccountMappingResponse(QBAccountMappingBase):
    id: uuid.UUID
    connection_id: uuid.UUID
    is_auto_created: bool
    created_at: datetime
    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Entity Mappings
# ---------------------------------------------------------------------------


class QBEntityMappingResponse(BaseModel):
    id: uuid.UUID
    entity_type: str
    pos_entity_id: uuid.UUID
    pos_entity_name: str
    qb_entity_id: str
    qb_entity_type: str
    qb_entity_name: str
    sync_direction: str
    last_synced_at: datetime | None = None
    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Sync
# ---------------------------------------------------------------------------


class QBSyncJobResponse(BaseModel):
    id: uuid.UUID
    job_type: str
    entity_type: str
    entity_id: uuid.UUID | None = None
    priority: int
    status: str
    error_message: str | None = None
    retry_count: int
    created_at: datetime
    completed_at: datetime | None = None
    processing_duration_ms: int | None = None
    model_config = {"from_attributes": True}


class QBSyncLogResponse(BaseModel):
    id: uuid.UUID
    sync_type: str
    pos_entity_type: str | None = None
    pos_entity_id: uuid.UUID | None = None
    qb_entity_type: str | None = None
    qb_entity_id: str | None = None
    action: str
    status: str
    error_message: str | None = None
    error_code: str | None = None
    duration_ms: int | None = None
    qb_doc_number: str | None = None
    amount_paisa: int | None = None
    created_at: datetime
    model_config = {"from_attributes": True}


class QBSyncStats(BaseModel):
    total_synced: int
    last_24h_synced: int
    last_24h_failed: int
    pending_jobs: int
    failed_jobs: int
    dead_letter_jobs: int
    last_sync_at: datetime | None = None
    sync_by_type: dict[str, int] = Field(default_factory=dict)


class QBManualSyncRequest(BaseModel):
    sync_type: str = Field(
        ...,
        pattern=r"^(full_sync|sync_items|sync_customers|sync_orders|daily_summary|setup_tax_codes)$",
    )
    date_from: datetime | None = None
    date_to: datetime | None = None
    entity_ids: list[uuid.UUID] | None = None


class QBSyncTriggerResponse(BaseModel):
    jobs_created: int
    message: str
    batch_id: uuid.UUID


# ---------------------------------------------------------------------------
# QB Account Discovery
# ---------------------------------------------------------------------------


class QBAccountInfo(BaseModel):
    id: str
    name: str
    account_type: str
    account_sub_type: str | None = None
    fully_qualified_name: str | None = None
    current_balance: float | None = None
    active: bool = True


class QBAccountListResponse(BaseModel):
    accounts: list[QBAccountInfo] = Field(default_factory=list)
    total: int


# ---------------------------------------------------------------------------
# Account Matching (Attempt 2 — replaces diagnostic/templates)
# ---------------------------------------------------------------------------


class QBMatchSignals(BaseModel):
    exact: float = 0.0
    anchor: float = 0.0
    jaccard: float = 0.0
    synonym: float = 0.0
    type_match: float = 0.0
    substring: float = 0.0


class QBMatchCandidate(BaseModel):
    qb_account_id: str
    qb_account_name: str
    qb_account_type: str
    qb_account_sub_type: str | None = None
    fully_qualified_name: str | None = None
    active: bool = True
    score: float = Field(..., ge=0.0, le=1.0)
    signals: QBMatchSignals
    confidence: str


class QBMatchItem(BaseModel):
    """One POS accounting need and its match status."""

    need_key: str
    need_label: str
    need_description: str
    expected_qb_types: list[str] = Field(default_factory=list)
    expected_qb_sub_type: str | None = None
    required: bool = False
    status: str  # matched, candidates, unmatched
    best_match: QBMatchCandidate | None = None
    candidates: list[QBMatchCandidate] = Field(default_factory=list)
    decision: str | None = None
    decision_account_id: str | None = None
    decision_account_name: str | None = None


class QBUnmappedAccount(BaseModel):
    qb_account_id: str
    qb_account_name: str
    qb_account_type: str
    qb_account_sub_type: str | None = None
    fully_qualified_name: str | None = None
    active: bool = True
    suggested_mapping_type: str | None = None


class QBDecisionSummary(BaseModel):
    use_existing: int = 0
    create_new: int = 0
    skip: int = 0
    pending: int = 0
    ready_to_apply: bool = False


class QBMatchResult(BaseModel):
    """Full match result — POS needs vs QB accounts."""

    id: str
    created_at: str
    is_live: bool = False
    total_needs: int
    total_qb_accounts: int
    matched: int
    candidates: int
    unmatched: int
    required_total: int = 0
    required_matched: int = 0
    coverage_pct: float
    health_grade: str
    items: list[QBMatchItem] = Field(default_factory=list)
    unmapped_qb_accounts: list[QBUnmappedAccount] = Field(default_factory=list)
    decision_summary: QBDecisionSummary | None = None
    apply_result: dict | None = None
    applied_at: str | None = None


class QBMatchResultSummary(BaseModel):
    id: str
    created_at: str
    health_grade: str
    matched: int
    candidates: int
    unmatched: int
    total_needs: int
    coverage_pct: float
    is_live: bool = False


class QBDecision(BaseModel):
    index: int = Field(..., ge=0)
    decision: str = Field(..., pattern=r"^(use_existing|create_new|skip)$")
    qb_account_id: str | None = None
    qb_account_name: str | None = None


class QBDecisionsRequest(BaseModel):
    decisions: list[QBDecision] = Field(..., min_length=1)


class QBMatchApplyResponse(BaseModel):
    accounts_created: int
    mappings_created: int
    skipped: int
    errors: list[str] = Field(default_factory=list)
    details: list[str] = Field(default_factory=list)
    result_id: str


# ---------------------------------------------------------------------------
# Health Check
# ---------------------------------------------------------------------------


class QBHealthCheckDetail(BaseModel):
    mapping_id: str
    mapping_type: str
    qb_account_id: str
    qb_account_name: str
    status: str
    issue: str | None = None
    message: str
    current_name: str | None = None


class QBHealthCheckResponse(BaseModel):
    grade: str
    total_mappings: int
    healthy: int
    warnings: int
    critical: int
    checked_at: str
    details: list[QBHealthCheckDetail] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# CoA Snapshots
# ---------------------------------------------------------------------------


class QBSnapshotSummary(BaseModel):
    """Summary view of a CoA snapshot (without full account data)."""

    id: uuid.UUID
    snapshot_type: str
    account_count: int
    is_locked: bool
    version: int
    qb_company_name: str
    qb_realm_id: str
    fetched_at: datetime
    created_at: datetime
    notes: str | None = None
    model_config = {"from_attributes": True}


class QBSnapshotDetail(QBSnapshotSummary):
    """Full snapshot including account data."""

    coa_data: list[dict] = Field(default_factory=list)


class QBSnapshotCreateResponse(BaseModel):
    """Response after creating snapshots."""

    backup_id: str
    working_copy_id: str
    account_count: int
    version: int
    company_name: str
    fetched_at: str
